[AIRFLOW-932] Do not mark tasks removed when backfilling

bolkedebruin opened the pull request:
In a backfill one can specify a specific task to execute. We
create a subset of the orginal tasks in a subdag from the original dag.
The subdag has the same name as the original dag. This breaks
the integrity check of a dag_run as tasks are suddenly not in
scope any more.

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-921

Testing Done:
- New unit tests added

@aoen @saguziel @jlowin 

note to self: I need to carry this over to AIRFLOW-910

aoen commented on airflow/jobs.py:
```diff
@@ -1761,7 +1762,9 @@ def get_task_instances_for_dag_run(dag_run):
                 tasks = {}
                 for ti in dag_run.get_task_instances():
-                    tasks[ti.key] = ti
+                    # the specified dag can be a subset of the whole dag (sub_dag)
```
Nit: Might change to something like the following to be more clear # Need to filter on dag.task_ids because the specified dag might be a subset of the whole DAG and might not encompass all the task instances in the dagrun

It would be kinda nice if we could rework this so this wasn't a special case (i.e. dag_run.get_task_instances() would only return the relevant task instances) but this would probably be too much work for now.

aoen commented on airflow/models.py:
```diff
@@ -3978,6 +3983,11 @@ def update_state(self, session=None):
 
         dag = self.get_dag()
+
+        # backfills create subsets of dags with the same name
```
I feel this is a bit of a hack and adds unnecessary complexity. I still think we could just "REMOVE" the task instances just in the scheduler.

@mistercrunch what do you think?

bolkedebruin replied:
I think the proper way of doing it, is not to create a subdag but let the backfill job (/ scheduler job in the future) figure this out properly. Ie. sub_dag should be renamed "subset" and we could just filter out what tasks to run from the whole dag.

bolkedebruin replied:
And the update_state can't work without this. "update_state" itself doesn't remove task instances, but it needs the whole DAG to assess state, which in turn will be communicated to the user on the frontend.

aoen commented on airflow/models.py:
```diff
@@ -4062,7 +4072,11 @@ def verify_integrity(self, session=None):
+                # tasks of a whole dag but with the same name. If that is the case
+                # we should not mark tasks removed as we might not consider
+                # the whole dag.
+                if self.state is not State.RUNNING and not self.is_backfill:
```
Ditto with the above comment.

mistercrunch replied:
It's a common use case to mutate the DAG locally in the hopes of generating a custom backfill. In those cases the user has the assumption that Airflow only knows about the tasks defined. The user would expect state changes only on tasks that are in scope.

bolkedebruin replied:
@mistercrunch we are not considering state changes of tasks here, but the state change of the whole dag run. The use case is clear to me, how to work it out on the DagRun level is bit different.

saguziel commented on airflow/models.py:
```diff
@@ -3918,6 +3924,9 @@ def get_task_instances(self, state=None, session=None):
                     tis = tis.filter(TI.state.in_(state))
 
+        if self.dag and self.dag.partial:
+            tis = tis.filter(TI.task_id.in_(self.dag.task_ids))
```
.with_hint(TI, 'USE INDEX (PRIMARY)', dialect_name='mysql')

mistercrunch replied:
[optional]

bolkedebruin committed:
AIRFLOW-932] Do not mark tasks removed when backfilling

In a backfill one can specify a specific task to execute. We
create a subset of the orginal tasks in a subdag from the original dag.
The subdag has the same name as the original dag. This breaks
the integrity check of a dag_run as tasks are suddenly not in
scope any more.
